core/src/event/contract_setup.rs

```diff
@@ -90,6 +94,66 @@ impl ContractInformation {
     }
 }
 
+#[derive(Clone)]
+pub struct NetworkTrace {
+    pub id: String,
+    pub network: String,
+    pub cached_provider: Arc<JsonRpcCachedProvider>,
+    pub start_block: Option<U64>,
+    pub end_block: Option<U64>,
+}
+
+impl NetworkTrace {
+    pub fn is_live_indexing(&self) -> bool {
+        self.end_block.is_none()
+    }
+}
+
+#[derive(Clone)]
+pub struct TraceInformation {
+    pub name: String,
+    pub details: Vec<NetworkTrace>,
+    pub reorg_safe_distance: bool,
+}
+
+impl TraceInformation {
+    pub fn create(
+        native_transfers: NativeTransfers,
+        network_providers: &[CreateNetworkProvider],
+    ) -> Result<TraceInformation, CreateContractInformationError> {
+        let mut details = vec![];
+        let trace_networks = native_transfers.networks.unwrap_or_default();
+
+        for n in &trace_networks {
+            let name = n.network.clone();
+            let provider = network_providers.iter().find(|item| item.network_name == name);
```

Comment:
now we have `network_providers.iter().find(|item| item.network_name == name)` twice lets do a nice helper again sharing the code as much as we can